File: fibonacci-number.py
class Solution:
    # O(n), O(1)
    def fib(self, n: int) -> int:
        x, y = 0, 1
        for _ in range(n):
            x, y = y, x + y
        
        return x

    # iteration, dynamic programming, tabulation, space complexity = O(n)
    def fib(self, n: int) -> int:
        F = [0, 1]
        
        for i in range(2, n + 1):
            F.append(F[i - 1] + F[i - 2])
        
        return F[n]
    
    # recursion, dynamic programming, memoization, space complexity = O(n)
    F = collections.defaultdict(int)
    def fib(self, n: int) -> int:
        if n <= 1:
            return n
        
        if self.F[n]:
            return self.F[n]
        
        self.F[n] = self.fib(n - 1) + self.fib(n - 2)
        
        return self.F[n]
    
    # recursion
    def fib(self, n: int) -> int:
        if n <= 1:
            return n

        return self.fib(n - 1) + self.fib(n - 2)

File: fibonacci-number.java
class Solution {
    public int fib(int n) {
        if(n == 0)
            return 0;
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        for(int i=2; i<=n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }
}

class Solution {
    
    int[] dp = new int[31];

    public int fib(int n) {
        if(n <= 1)
            return n;
        
        if(dp[n] > 0)
            return dp[n];
        else
            return dp[n] = fib(n - 1) + fib(n - 2);
    }
}

class Solution {
    public int fib(int n) {
        int x = 0, y = 1;
        for(int i=0; i<n; i++) {
            int temp = x;
            x = y;
            y += temp;
        }
        return x;
    }
}